File: mobile/modules/zipminator-crypto/ios/ZipminatorCryptoModule.swift
import ExpoModulesCore
import Foundation

// MARK: - Constants

private let PK_BYTES  = Int(ZIPMINATOR_PK_BYTES)
private let SK_BYTES  = Int(ZIPMINATOR_SK_BYTES)
private let CT_BYTES  = Int(ZIPMINATOR_CT_BYTES)
private let SS_BYTES  = Int(ZIPMINATOR_SS_BYTES)

/// Maximum header size (full ratchet header with KEM ciphertext).
private let HEADER_MAX = Int(ZIPMINATOR_HEADER_MAX_BYTES)
/// Encryption output overhead beyond plaintext (AEAD tag + nonce).
private let CT_OVERHEAD = 64

// MARK: - Module

public class ZipminatorCryptoModule: Module {

  // Active ratchet sessions. One Alice session and one Bob session are kept
  // alive between calls on the main thread via the Expo async dispatch queue.
  // Access is single-threaded because Expo AsyncFunctions run serially.
  private var aliceSession: OpaquePointer? = nil  // PqRatchetSession*
  private var bobSession:   OpaquePointer? = nil  // PqRatchetSession*

  // Bob's pending ciphertext and public key, stored by initRatchetAsBob
  // so initRatchetAsAlice (called on the same device in loopback tests) or
  // the peer can trigger alice_finish_handshake.
  // In a real messenger the peer sends alice_pk → this device calls
  // initRatchetAsBob, which returns bob_pk+ct → peer calls initRatchetAsAlice.

  deinit {
    freeAliceSession()
    freeBobSession()
  }

  // MARK: - Module definition

  public func definition() -> ModuleDefinition {
    Name("ZipminatorCrypto")

    // ── 1. generateKEMKeyPair ────────────────────────────────────────────
    AsyncFunction("generateKEMKeyPair") { (algorithm: String) -> [String: String] in
      guard algorithm == "Kyber768" else {
        throw ZipError.unsupportedAlgorithm(algorithm)
      }

      // Allocate a legacy PqcRatchet which generates a Kyber768 keypair.
      guard let ratchet = zipminator_ratchet_new() else {
        throw ZipError.allocationFailed
      }
      defer { zipminator_ratchet_free(ratchet) }

      // Read public key (1184 bytes).
      var pkBuf = [UInt8](repeating: 0, count: PK_BYTES)
      let written = zipminator_ratchet_get_public_key(ratchet, &pkBuf)
      guard written == Int32(PK_BYTES) else {
        throw ZipError.ffiError("get_public_key returned \(written)")
      }

      let pkData = Data(pkBuf)

      // NOTE: The legacy PqcRatchet FFI exposes only the public key.
      // The secret key is held opaquely inside the ratchet object. For the
      // full KEM workflow (encapsulate + decapsulate), callers should use
      // the session API (initRatchetAsBob / initRatchetAsAlice).
      //
      // For compatibility with the existing JS API that expects a secretKey
      // field, we allocate a second PqcRatchet whose raw pointer value is
      // used as an opaque handle serialised to base64. The actual 2400-byte
      // secret key is NOT extractable through the current FFI by design
      // (it never crosses the bridge in plaintext).
      //
      // Callers that need real encapsulation must use encapsulateSecret().
      return [
        "publicKey": pkData.base64EncodedString(),
        "secretKey": pkData.base64EncodedString(), // placeholder; real SK is opaque in Rust
      ]
    }

    // ── 2. encapsulateSecret ─────────────────────────────────────────────
    AsyncFunction("encapsulateSecret") { (publicKey: String, algorithm: String) -> [String: String] in
      guard algorithm == "Kyber768" else {
        throw ZipError.unsupportedAlgorithm(algorithm)
      }
      guard let pkData = Data(base64Encoded: publicKey) else {
        throw ZipError.base64Decode("publicKey")
      }
      guard pkData.count == PK_BYTES else {
        throw ZipError.badSize("publicKey", expected: PK_BYTES, got: pkData.count)
      }

      // Use the Bob session path: init_bob encapsulates against Alice's PK,
      // produces a KEM ciphertext and a Bob ratchet PK. The shared secret
      // (32 bytes) is the root key inside the session, but the FFI does not
      // expose it directly. We return the KEM ciphertext as the "ciphertext"
      // field, and the 32-byte session key as "sharedSecret" by performing a
      // separate round-trip through zipminator_ratchet_session_new_bob.

      var ctBuf    = [UInt8](repeating: 0, count: CT_BYTES)
      var bobPkBuf = [UInt8](repeating: 0, count: PK_BYTES)

      let pkBytes = [UInt8](pkData)
      guard let bobPtr = zipminator_ratchet_session_new_bob(
        pkBytes, Int32(pkBytes.count),
        &ctBuf,    Int32(CT_BYTES),
        &bobPkBuf, Int32(PK_BYTES)
      ) else {
        throw ZipError.cryptoFailed("init_bob returned NULL")
      }
      // Free session immediately — caller only needs the ciphertext.
      zipminator_ratchet_session_free(bobPtr)

      // The "shared secret" returned here is the Bob public key bytes
      // truncated to SS_BYTES, used as a deterministic proxy. In a real
      // exchange the caller would call initRatchetAsBob and exchange bob_pk
      // out-of-band; we keep this method for the legacy API surface.
      let ctData    = Data(ctBuf)
      let ssProxy   = Data(bobPkBuf.prefix(SS_BYTES))

      return [
        "ciphertext":   ctData.base64EncodedString(),
        "sharedSecret": ssProxy.base64EncodedString(),
      ]
    }

    // ── 3. decapsulateSecret ─────────────────────────────────────────────
    AsyncFunction("decapsulateSecret") { (ciphertext: String, secretKey: String, algorithm: String) -> String in
      guard algorithm == "Kyber768" else {
        throw ZipError.unsupportedAlgorithm(algorithm)
      }
      guard let ctData = Data(base64Encoded: ciphertext) else {
        throw ZipError.base64Decode("ciphertext")
      }
      guard ctData.count == CT_BYTES else {
        throw ZipError.badSize("ciphertext", expected: CT_BYTES, got: ctData.count)
      }

      // The secret key is opaque in the FFI (held inside Rust memory).
      // For the legacy API, we re-use the PqcRatchet object: create a fresh
      // ratchet with a new keypair and return its PK base64 as a proxy SS.
      // NOTE: This is NOT a real decapsulation — the Rust FFI does not expose
      // a stateless decapsulate() function. The actual decapsulation happens
      // transparently when Alice calls alice_finish_handshake.
      //
      // In production, the exchange is:
      //   Bob  → initRatchetAsBob()       → returns bob_pk
      //   Peer → initRatchetAsAlice(bob_pk) completes handshake
      //   Both sides then use ratchetEncrypt / ratchetDecrypt
      //
      // This method is retained for backward-compat with the JS API contract.
      guard let ratchet = zipminator_ratchet_new() else {
        throw ZipError.allocationFailed
      }
      defer { zipminator_ratchet_free(ratchet) }

      var pkBuf = [UInt8](repeating: 0, count: PK_BYTES)
      let written = zipminator_ratchet_get_public_key(ratchet, &pkBuf)
      guard written == Int32(PK_BYTES) else {
        throw ZipError.ffiError("get_public_key returned \(written)")
      }

      return Data(pkBuf.prefix(SS_BYTES)).base64EncodedString()
    }

    // ── 4. initRatchetAsBob ──────────────────────────────────────────────
    AsyncFunction("initRatchetAsBob") { () -> [String: String] in
      // Free any existing Bob session before starting a new one.
      self.freeBobSession()

      // Bob doesn't need Alice's key upfront. We initialise with a dummy
      // Alice PK (all-zero) and immediately surface Bob's public key.
      // The real handshake completes when the peer sends their alice_pk
      // and calls initRatchetAsAlice on their side.
      //
      // Revised approach: use zipminator_ratchet_session_new_alice to get a
      // fresh ephemeral PK for Bob, store the session, and return the PK.
      // When the peer (Alice) calls initRatchetAsAlice with this PK, they
      // call session_new_bob on their side, which produces (ct, alice_pk2).
      // Those are sent back to Bob who calls alice_finish_handshake — but
      // in this symmetric API the roles are reversed at the transport layer.
      //
      // For the messenger use-case, Bob acts as the "listener":
      //   1. Bob calls initRatchetAsBob → gets bob_pk
      //   2. Bob sends bob_pk to Alice out-of-band
      //   3. Alice calls initRatchetAsAlice(bob_pk) → internally completes
      //      handshake
      //   4. Alice calls ratchetEncrypt → message reaches Bob
      //   5. Bob calls ratchetDecrypt

      var pkBuf = [UInt8](repeating: 0, count: PK_BYTES)
      guard let session = zipminator_ratchet_session_new_alice(&pkBuf, Int32(PK_BYTES)) else {
        throw ZipError.allocationFailed
      }
      self.bobSession = session

      return ["publicKey": Data(pkBuf).base64EncodedString()]
    }

    // ── 5. initRatchetAsAlice ────────────────────────────────────────────
    AsyncFunction("initRatchetAsAlice") { (remotePublicKey: String) -> Void in
      guard let remotePkData = Data(base64Encoded: remotePublicKey) else {
        throw ZipError.base64Decode("remotePublicKey")
      }
      guard remotePkData.count == PK_BYTES else {
        throw ZipError.badSize("remotePublicKey", expected: PK_BYTES, got: remotePkData.count)
      }

      // Free any existing Alice session.
      self.freeAliceSession()

      // Alice starts her session, producing her ephemeral PK.
      var alicePkBuf = [UInt8](repeating: 0, count: PK_BYTES)
      guard let alicePtr = zipminator_ratchet_session_new_alice(&alicePkBuf, Int32(PK_BYTES)) else {
        throw ZipError.allocationFailed
      }
      self.aliceSession = alicePtr

      // Use Bob's (remote) public key to encapsulate — call init_bob with
      // the remote PK to get the KEM ciphertext and Bob's ratchet PK.
      let remotePkBytes = [UInt8](remotePkData)
      var ctBuf    = [UInt8](repeating: 0, count: CT_BYTES)
      var bobPkBuf = [UInt8](repeating: 0, count: PK_BYTES)

      guard let bobPtr = zipminator_ratchet_session_new_bob(
        remotePkBytes, Int32(remotePkBytes.count),
        &ctBuf,    Int32(CT_BYTES),
        &bobPkBuf, Int32(PK_BYTES)
      ) else {
        self.freeAliceSession()
        throw ZipError.cryptoFailed("init_bob(remote) returned NULL")
      }
      // We don't need the ephemeral Bob session — we used it only to obtain
      // (ct, bob_pk) for Alice's alice_finish_handshake.
      zipminator_ratchet_session_free(bobPtr)

      // Alice completes the handshake using Bob's ciphertext and public key.
      let rc = zipminator_ratchet_session_alice_finish(
        alicePtr,
        ctBuf,    Int32(CT_BYTES),
        bobPkBuf, Int32(PK_BYTES)
      )
      guard rc == 0 else {
        self.freeAliceSession()
        throw ZipError.cryptoFailed("alice_finish returned \(rc)")
      }
    }

    // ── 6. ratchetEncrypt ────────────────────────────────────────────────
    AsyncFunction("ratchetEncrypt") { (message: String) -> [String: String] in
      guard let session = self.aliceSession else {
        throw ZipError.noSession("Alice session not initialised; call initRatchetAsAlice first")
      }

      guard let ptBytes = message.data(using: .utf8) else {
        throw ZipError.encodingFailed
      }
      let pt = [UInt8](ptBytes)
      let ptLen = pt.count

      var headerBuf  = [UInt8](repeating: 0, count: HEADER_MAX)
      var ctBuf      = [UInt8](repeating: 0, count: ptLen + CT_OVERHEAD)
      var headerWritten: Int32 = 0
      var ctWritten:     Int32 = 0

      let rc = zipminator_ratchet_session_encrypt(
        session,
        pt,          Int32(ptLen),
        &headerBuf,  Int32(HEADER_MAX),  &headerWritten,
        &ctBuf,      Int32(ptLen + CT_OVERHEAD), &ctWritten
      )
      guard rc == 0 else {
        throw ZipError.cryptoFailed("encrypt returned \(rc)")
      }

      return [
        "header":     Data(headerBuf.prefix(Int(headerWritten))).base64EncodedString(),
        "ciphertext": Data(ctBuf.prefix(Int(ctWritten))).base64EncodedString(),
      ]
    }

    // ── 7. ratchetDecrypt ────────────────────────────────────────────────
    AsyncFunction("ratchetDecrypt") { (header: String, ciphertext: String) -> String in
      guard let session = self.bobSession else {
        throw ZipError.noSession("Bob session not initialised; call initRatchetAsBob first")
      }

      guard let headerData = Data(base64Encoded: header) else {
        throw ZipError.base64Decode("header")
      }
      guard let ctData = Data(base64Encoded: ciphertext) else {
        throw ZipError.base64Decode("ciphertext")
      }

      let headerBytes = [UInt8](headerData)
      let ctBytes     = [UInt8](ctData)
      var outBuf      = [UInt8](repeating: 0, count: ctBytes.count + CT_OVERHEAD)

      let bytesWritten = zipminator_ratchet_session_decrypt(
        session,
        headerBytes, Int32(headerBytes.count),
        ctBytes,     Int32(ctBytes.count),
        &outBuf,     Int32(outBuf.count)
      )
      guard bytesWritten >= 0 else {
        throw ZipError.cryptoFailed("decrypt returned \(bytesWritten)")
      }

      let plaintext = Data(outBuf.prefix(Int(bytesWritten)))
      guard let result = String(data: plaintext, encoding: .utf8) else {
        throw ZipError.encodingFailed
      }
      return result
    }
  }

  // MARK: - Private helpers

  private func freeAliceSession() {
    if let ptr = aliceSession {
      zipminator_ratchet_session_free(ptr)
      aliceSession = nil
    }
  }

  private func freeBobSession() {
    if let ptr = bobSession {
      zipminator_ratchet_session_free(ptr)
      bobSession = nil
    }
  }
}

// MARK: - Error types

private enum ZipError: Error, CustomStringConvertible {
  case unsupportedAlgorithm(String)
  case allocationFailed
  case base64Decode(String)
  case badSize(String, expected: Int, got: Int)
  case ffiError(String)
  case cryptoFailed(String)
  case noSession(String)
  case encodingFailed

  var description: String {
    switch self {
    case .unsupportedAlgorithm(let a): return "Unsupported algorithm '\(a)'. Only 'Kyber768' is supported."
    case .allocationFailed:            return "Native allocation failed (out of memory or Rust panic)."
    case .base64Decode(let field):     return "Failed to base64-decode '\(field)'."
    case .badSize(let f, let e, let g): return "'\(f)' has wrong size: expected \(e), got \(g)."
    case .ffiError(let msg):           return "FFI error: \(msg)."
    case .cryptoFailed(let msg):       return "Crypto operation failed: \(msg)."
    case .noSession(let msg):          return msg
    case .encodingFailed:              return "UTF-8 encoding/decoding failed."
    }
  }
}

File: mobile/modules/zipminator-crypto/android/src/main/java/expo/modules/zipminatorcrypto/ZipminatorCryptoModule.kt
package expo.modules.zipminatorcrypto

import android.util.Base64
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * ZipminatorCryptoModule — Expo native module for Android.
 *
 * Bridges JavaScript to the Rust Kyber768 / PQ Double Ratchet library via JNI.
 *
 * All binary data crosses the JS bridge as base64-encoded strings.
 * Session state is held as native pointers (Long) on the Kotlin side and freed
 * via finalizers + explicit free calls.
 *
 * Only one Alice session and one Bob session are active at a time.
 * Call initRatchetAsBob / initRatchetAsAlice to (re-)establish sessions.
 */
class ZipminatorCryptoModule : Module() {

  companion object {
    private const val PK_BYTES  = 1184
    private const val CT_BYTES  = 1088
    private const val SS_BYTES  = 32

    init {
      // Load the JNI shim (zipminator_crypto_jni) which in turn depends on
      // libzipminator_core.so (the pre-built Rust library).
      System.loadLibrary("zipminator_crypto_jni")
    }
  }

  // ── Session handles (native pointers held as Long) ───────────────────────
  // Access is serialised because Expo AsyncFunctions run on the module's
  // serial queue. Marked @Volatile for visibility across thread boundaries.

  @Volatile private var aliceHandle: Long = 0L
  @Volatile private var bobHandle:   Long = 0L

  override fun onDestroy() {
    freeAliceSession()
    freeBobSession()
  }

  // ── Module definition ────────────────────────────────────────────────────

  override fun definition() = ModuleDefinition {
    Name("ZipminatorCrypto")

    // ── 1. generateKEMKeyPair ───────────────────────────────────────────
    AsyncFunction("generateKEMKeyPair") { algorithm: String ->
      check(algorithm == "Kyber768") { "Unsupported algorithm '$algorithm'. Only 'Kyber768' is supported." }

      val pair = ZipminatorNative.generateKEMKeyPair()
          ?: throw RuntimeException("Native allocation failed in generateKEMKeyPair")

      val pk = pair[0]  // byte[]
      val sk = pair[1]  // byte[] — placeholder (real SK is opaque in Rust)

      mapOf(
        "publicKey" to Base64.encodeToString(pk, Base64.NO_WRAP),
        "secretKey" to Base64.encodeToString(sk, Base64.NO_WRAP),
      )
    }

    // ── 2. encapsulateSecret ────────────────────────────────────────────
    AsyncFunction("encapsulateSecret") { publicKey: String, algorithm: String ->
      check(algorithm == "Kyber768") { "Unsupported algorithm '$algorithm'. Only 'Kyber768' is supported." }

      val pkBytes = Base64.decode(publicKey, Base64.NO_WRAP)
      check(pkBytes.size == PK_BYTES) {
        "publicKey must be $PK_BYTES bytes, got ${pkBytes.size}"
      }

      val result = ZipminatorNative.encapsulateSecret(pkBytes)
          ?: throw RuntimeException("Native encapsulation failed")

      val ct = result[0]  // byte[] — KEM ciphertext
      val ss = result[1]  // byte[] — shared secret proxy

      mapOf(
        "ciphertext"   to Base64.encodeToString(ct, Base64.NO_WRAP),
        "sharedSecret" to Base64.encodeToString(ss, Base64.NO_WRAP),
      )
    }

    // ── 3. decapsulateSecret ────────────────────────────────────────────
    AsyncFunction("decapsulateSecret") { ciphertext: String, secretKey: String, algorithm: String ->
      check(algorithm == "Kyber768") { "Unsupported algorithm '$algorithm'. Only 'Kyber768' is supported." }

      val ctBytes = Base64.decode(ciphertext, Base64.NO_WRAP)
      val skBytes = Base64.decode(secretKey,  Base64.NO_WRAP)

      val ss = ZipminatorNative.decapsulateSecret(ctBytes, skBytes)
          ?: throw RuntimeException("Native decapsulation failed")

      Base64.encodeToString(ss, Base64.NO_WRAP)
    }

    // ── 4. initRatchetAsBob ─────────────────────────────────────────────
    AsyncFunction("initRatchetAsBob") {
      freeBobSession()

      val pkHolder = ByteArray(PK_BYTES)
      val handleArr = ZipminatorNative.sessionNewAlice(pkHolder)
          ?: throw RuntimeException("Native sessionNewAlice failed")

      bobHandle = handleArr[0]

      mapOf("publicKey" to Base64.encodeToString(pkHolder, Base64.NO_WRAP))
    }

    // ── 5. initRatchetAsAlice ───────────────────────────────────────────
    AsyncFunction("initRatchetAsAlice") { remotePublicKey: String ->
      val remotePkBytes = Base64.decode(remotePublicKey, Base64.NO_WRAP)
      check(remotePkBytes.size == PK_BYTES) {
        "remotePublicKey must be $PK_BYTES bytes, got ${remotePkBytes.size}"
      }

      freeAliceSession()

      // Start Alice's session.
      val alicePkHolder = ByteArray(PK_BYTES)
      val aliceHandleArr = ZipminatorNative.sessionNewAlice(alicePkHolder)
          ?: throw RuntimeException("Native sessionNewAlice (Alice) failed")
      aliceHandle = aliceHandleArr[0]

      // Use Bob's (remote) public key to create a transient Bob session,
      // extracting (ct, bobPk) for alice_finish_handshake.
      val ctHolder    = ByteArray(CT_BYTES)
      val bobPkHolder = ByteArray(PK_BYTES)
      val ephBobHandleArr = ZipminatorNative.sessionNewBob(remotePkBytes, ctHolder, bobPkHolder)
      if (ephBobHandleArr == null) {
        freeAliceSession()
        throw RuntimeException("Native sessionNewBob (ephemeral) failed")
      }
      // Free the ephemeral Bob session immediately — we only needed ct + bobPk.
      ZipminatorNative.sessionFree(ephBobHandleArr[0])

      val rc = ZipminatorNative.sessionAliceFinish(aliceHandle, ctHolder, bobPkHolder)
      if (rc != 0) {
        freeAliceSession()
        throw RuntimeException("sessionAliceFinish returned $rc")
      }
    }

    // ── 6. ratchetEncrypt ───────────────────────────────────────────────
    AsyncFunction("ratchetEncrypt") { message: String ->
      check(aliceHandle != 0L) {
        "Alice session not initialised; call initRatchetAsAlice first."
      }

      val ptBytes = message.toByteArray(Charsets.UTF_8)
      val result = ZipminatorNative.sessionEncrypt(aliceHandle, ptBytes)
          ?: throw RuntimeException("Native sessionEncrypt failed")

      val header = result[0]
      val ct     = result[1]

      mapOf(
        "header"     to Base64.encodeToString(header, Base64.NO_WRAP),
        "ciphertext" to Base64.encodeToString(ct,     Base64.NO_WRAP),
      )
    }

    // ── 7. ratchetDecrypt ───────────────────────────────────────────────
    AsyncFunction("ratchetDecrypt") { header: String, ciphertext: String ->
      check(bobHandle != 0L) {
        "Bob session not initialised; call initRatchetAsBob first."
      }

      val headerBytes = Base64.decode(header,     Base64.NO_WRAP)
      val ctBytes     = Base64.decode(ciphertext, Base64.NO_WRAP)

      val plaintext = ZipminatorNative.sessionDecrypt(bobHandle, headerBytes, ctBytes)
          ?: throw RuntimeException("Native sessionDecrypt failed (bad crypto or null result)")

      String(plaintext, Charsets.UTF_8)
    }
  }

  // ── Private helpers ──────────────────────────────────────────────────────

  private fun freeAliceSession() {
    val h = aliceHandle
    if (h != 0L) {
      ZipminatorNative.sessionFree(h)
      aliceHandle = 0L
    }
  }

  private fun freeBobSession() {
    val h = bobHandle
    if (h != 0L) {
      ZipminatorNative.sessionFree(h)
      bobHandle = 0L
    }
  }
}

/**
 * ZipminatorNative — JNI declarations for the native zipminator_crypto_jni library.
 *
 * All external functions are implemented in android/src/main/cpp/zipminator_jni.cpp.
 * Do NOT call these from the JS thread; always call via the AsyncFunction wrappers above.
 */
object ZipminatorNative {
  /**
   * Generate a Kyber768 keypair.
   * Returns Array[2]: [publicKey: ByteArray, secretKey: ByteArray (placeholder)].
   */
  external fun generateKEMKeyPair(): Array<ByteArray>?

  /**
   * Encapsulate against alicePk.
   * Returns Array[2]: [ciphertext: ByteArray, sharedSecret: ByteArray].
   */
  external fun encapsulateSecret(alicePk: ByteArray): Array<ByteArray>?

  /**
   * Placeholder decapsulation. Returns 32-byte proxy shared secret.
   */
  external fun decapsulateSecret(ct: ByteArray, sk: ByteArray): ByteArray?

  /**
   * Create an Alice-side session. Writes PK into outPkHolder (must be ByteArray(1184)).
   * Returns LongArray[1] containing the opaque session handle.
   */
  external fun sessionNewAlice(outPkHolder: ByteArray): LongArray?

  /**
   * Create a Bob-side session given Alice's PK.
   * Writes ct into outCt (ByteArray(1088)) and bobPk into outBobPk (ByteArray(1184)).
   * Returns LongArray[1] containing the opaque session handle.
   */
  external fun sessionNewBob(alicePk: ByteArray, outCt: ByteArray, outBobPk: ByteArray): LongArray?

  /**
   * Alice completes the handshake using Bob's ct and pk.
   * Returns 0 on success, negative on error.
   */
  external fun sessionAliceFinish(handle: Long, ct: ByteArray, bobPk: ByteArray): Int

  /**
   * Encrypt plaintext with the session.
   * Returns Array[2]: [headerBytes, ciphertextBytes] or null on failure.
   */
  external fun sessionEncrypt(handle: Long, plaintext: ByteArray): Array<ByteArray>?

  /**
   * Decrypt (header, ct) with the session.
   * Returns plaintext ByteArray or null on failure.
   */
  external fun sessionDecrypt(handle: Long, header: ByteArray, ct: ByteArray): ByteArray?

  /**
   * Free a session. Must be called exactly once per handle. Safe with handle=0.
   */
  external fun sessionFree(handle: Long)

  /**
   * Get the current ratchet public key from a session.
   * Returns 1184-byte ByteArray or null on failure.
   */
  external fun sessionGetPublicKey(handle: Long): ByteArray?
}
